presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java

```diff
@@ -689,31 +687,103 @@ protected ExtractionResult visitNullLiteral(NullLiteral node, Boolean complement
         Object left = ExpressionInterpreter.expressionOptimizer(comparison.getLeft(), metadata, session, expressionTypes).optimize(NoOpSymbolResolver.INSTANCE);
         Object right = ExpressionInterpreter.expressionOptimizer(comparison.getRight(), metadata, session, expressionTypes).optimize(NoOpSymbolResolver.INSTANCE);
 
-        if (left instanceof SymbolReference && !(right instanceof Expression)) {
-            return Optional.of(new NormalizedSimpleComparison((SymbolReference) left, comparison.getType(), new NullableValue(expressionTypes.get(comparison.getRight()), right)));
+        Type leftType = expressionTypes.get(comparison.getLeft());
+        Type rightType = expressionTypes.get(comparison.getRight());
+        Preconditions.checkArgument(leftType.equals(rightType), "left and right type do not match in comparison expression (%s)", comparison);
+
+        if (left instanceof Expression && right instanceof Expression) {
+            return Optional.empty();
+        }
+
+        if (!(left instanceof Expression) && !(right instanceof Expression)) {
+            return Optional.empty();
+        }
+
+        Expression symbolExpression;
+        ComparisonExpressionType comparisonType;
+        NullableValue value;
+
+        if (left instanceof Expression) {
+            symbolExpression = comparison.getLeft();
+            comparisonType = comparison.getType();
+            value = new NullableValue(rightType, right);
+        }
+        else {
+            symbolExpression = comparison.getRight();
+            comparisonType = comparison.getType().flip();
+            value = new NullableValue(leftType, left);
+        }
+
+        if (!checkSymbolExpressionValid(metadata, expressionTypes, symbolExpression)) {
+            return Optional.empty();
         }
-        if (right instanceof SymbolReference && !(left instanceof Expression)) {
-            return Optional.of(new NormalizedSimpleComparison((SymbolReference) right, comparison.getType().flip(), new NullableValue(expressionTypes.get(comparison.getLeft()), left)));
+
+        return Optional.of(new NormalizedSimpleComparison(symbolExpression, comparisonType, value));
+    }
+
+    private static Type typeOf(Expression expression, Session session, Metadata metadata, Map<Symbol, Type> types)
+    {
+        IdentityHashMap<Expression, Type> expressionTypes = ExpressionAnalyzer.getExpressionTypes(session, metadata, new SqlParser(), types, expression, emptyList() /* parameters already replaced */);
+        return expressionTypes.get(expression);
+    }
+
+    private static boolean checkSymbolExpressionValid(Metadata metadata, IdentityHashMap<Expression, Type> expressionTypes, Expression symbolExpression)
```

Comment:
Maybe change the name to better express what this method does. What it actually does - it just checks whether the cast chain is strictly widening. Maybe call it `downcastsPresent` and invert the result?